torch/csrc/autograd/variable.h

```diff
@@ -587,172 +367,44 @@ inline const Variable& as_variable_ref(const at::Tensor& tensor) {
-    /*version_counter=*/0,
-    /*allow_tensor_metadata_change=*/false);
-  self_impl_copy->set_autograd_meta(c10::guts::make_unique<AutogradMeta>(self_impl_copy.get(), false));
-  return at::Tensor(self_impl_copy);
-}
-
-// Gradient Node and Edges
-//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-
-inline Node* Variable::grad_fn_unsafe() const {
-  return get_autograd_meta()->grad_fn_.get();
-}
-
-inline void Variable::set_grad_accumulator(
-    std::weak_ptr<Node> grad_accumulator) {
-  get_autograd_meta()->grad_accumulator_ = std::move(grad_accumulator);
-}
-
-inline std::shared_ptr<Node> Variable::try_get_grad_accumulator() const {
-  return get_autograd_meta()->grad_accumulator_.lock();
-}
-
-inline Variable Variable::detach() const {
-  auto var = make_variable_view(*this, *this, /*is_differentiable=*/false, /*allow_tensor_metadata_change=*/false, Edge());
-#ifdef BUILD_NAMEDTENSOR
-  at::namedinference::propagate_names(var, *this);
-#endif
-  return var;
-}
-
-inline void Variable::set_gradient_edge(Edge edge) noexcept {
-  get_autograd_meta()->grad_fn_ = std::move(edge.function);
-  get_autograd_meta()->output_nr_ = edge.input_nr;
-}
-
-inline uint32_t Variable::output_nr() const noexcept {
-  return get_autograd_meta()->output_nr_;
-}
-
-inline bool Variable::is_leaf() const noexcept {
-  return get_autograd_meta()->grad_fn_ == nullptr;
-}
-
-// Versions
-//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-
-inline void Variable::set_version_counter(
-    const c10::VariableVersion& version_counter) noexcept {
-  unsafeGetTensorImpl()->set_version_counter(version_counter);
-}
-
-inline void Variable::bump_version() noexcept {
-  unsafeGetTensorImpl()->bump_version();
-}
-
-inline uint32_t Variable::current_version() const noexcept {
-  return unsafeGetTensorImpl()->version_counter().current_version();
-}
-
-inline const c10::VariableVersion& Variable::version_counter() const noexcept {
-  return unsafeGetTensorImpl()->version_counter();
-}
-
-// Hooks
-//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-
-inline void Variable::add_hook(std::shared_ptr<FunctionPreHook> hook) {
-  get_autograd_meta()->hooks_.push_back(std::move(hook));
-}
+}} // namespace torch::autograd
 
-inline const std::vector<std::shared_ptr<FunctionPreHook>>& Variable::hooks()
-    const noexcept {
-  return get_autograd_meta()->hooks_;
-}
+// NB: These template definitions shouldn't live here, but they require access
+// to the structure of AutogradMeta, so for now they live here.  Once
+// AutogradMeta is available inline, move these back to Tensor.h
+namespace at {
 
-inline void Variable::clear_hooks() {
-  get_autograd_meta()->hooks_.clear();
-}
+void _create_cpp_hook(const Tensor& self);
 
 template <typename T>
-auto Variable::register_hook(T&& hook) -> Variable::hook_return_void_t<T> {
+auto Tensor::register_hook(T&& hook) const -> Tensor::hook_return_void_t<T> {
```

Comment:
You define `using Variable = at::Tensor` earlier in this file. Why change them all here?

Reply:
Accident. I started off by moving all methods to Tensor (and therefore renamed them), and then later I added the alias (making the old definition variable). It's better for them to directly be based on Tensor though.